enterprise/server/raft/rangelease/rangelease.go

```diff
@@ -298,50 +307,61 @@ func (l *Lease) renewLeaseUntilValid(ctx context.Context) (returnedRecord *rfpb.
 		span.End()
 	}()
 	l.mu.Lock()
-	defer l.mu.Unlock()
 
 	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
 	defer cancel()
 
 	for {
 		select {
 		case <-ctx.Done():
+			l.log.Info("renewLeaseAUntilValid cancelled")
+			l.mu.Unlock()
 			return nil, ctx.Err()
 		default: // continue with for loop
 		}
 		if err := l.renewLease(ctx); err != nil {
+			l.mu.Unlock()
 			return nil, status.InternalErrorf("failed to renew lease: %s", err)
 		}
 		if err := l.verifyLease(ctx, l.leaseRecord); err == nil {
 			break
 		}
 	}
 
+	leaseRecord := l.leaseRecord
+	l.mu.Unlock()
+
+	l.stopMutex.Lock()
+	defer l.stopMutex.Unlock()
+
 	// We just renewed the lease. If there isn't already a background
 	// thread running to keep it renewed, start one now.
 	if l.stopped {
```

Comment:
In Java you could use double-checked locking here to avoid the overhead if a acquiring the lock in every pass through this method, but I think that the Go memory model doesn't allow this.